R/undesirable_operator_linter.R

```diff
@@ -19,33 +9,59 @@ op_types <- c(
 #' @seealso [linters] for a complete list of linters available in lintr.
 #' @export
 undesirable_operator_linter <- function(op = default_undesirable_operators) {
+  undesirable_operator_metadata <- merge(
+    # infix must be handled individually below; non-assignment `=` are always OK
+    infix_metadata[
+      infix_metadata$string_value != "%%" & !infix_metadata$xml_tag %in% c("EQ_SUB", "EQ_FORMALS"),
+    ],
+    infix_overload,
+    by = "xml_tag", all.x = TRUE
+  )
+
+  included_operators <- undesirable_operator_metadata$string_value %in% names(op) |
+    undesirable_operator_metadata$exact_string_value %in% names(op)
+  operator_nodes <- undesirable_operator_metadata$xml_tag[included_operators]
+  needs_exact_string <- !is.na(undesirable_operator_metadata$exact_string_value[included_operators])
+  operator_nodes[needs_exact_string] <- sprintf(
+    "%s[text() = '%s']",
+    operator_nodes[needs_exact_string],
+    undesirable_operator_metadata$exact_string_value[included_operators][needs_exact_string]
+  )
+
+  is_infix <- startsWith(names(op), "%")
+  if (any(is_infix)) {
+    operator_nodes <- c(operator_nodes, sprintf("SPECIAL[text() = '%s']", names(op)[is_infix]))
+  }
+
+  if (length(operator_nodes) == 0L) {
+    stop("Did not recognize any valid operators in request for: ", toString(names(op)))
+  }
+
+  operator_nodes <- paste0("self::", operator_nodes)
+
   Linter(function(source_expression) {
-    lapply(
-      ids_with_token(source_expression, op_types, fun = `%in%`),
-      function(id) {
-        token <- with_id(source_expression, id)
-        op_name <- token[["text"]]
-        if (op_name %in% names(op)) {
-          line_num <- token[["line1"]]
-          start_col_num <- token[["col1"]]
-          end_col_num <- token[["col2"]]
-          msg <- sprintf("Operator `%s` is undesirable.", op_name)
-          alt_op <- op[[op_name]]
-          if (!is.na(alt_op)) {
-            msg <- paste(msg, alt_op)
-          }
-
-          Lint(
-            filename = source_expression[["filename"]],
-            line_number = line_num,
-            column_number = start_col_num,
-            type = "warning",
-            message = paste0(msg, collapse = " "),
-            line = source_expression[["lines"]][[as.character(line_num)]],
-            ranges = list(c(start_col_num, end_col_num))
-          )
+    if (!is_lint_level(source_expression, "expression")) {
+      return(list())
+    }
+
+    xml <- source_expression$xml_parsed_content
+
+    xpath <- glue::glue("//*[ { xp_or(operator_nodes) } ]")
```

Comment:
Could be moved into the factory as well, since that's where most of the XPath is constructed.

Reply:
I would keep it here... I think it's clearer/ more consistent w other linters

Reply:
it might be a separate issue to tackle for performance more globally.

will definitely change to use |